pkg/action/install.go

```diff
@@ -103,6 +103,50 @@ func NewInstall(cfg *Configuration) *Install {
 	}
 }
 
+func (i *Install) installCRDs(crds []*chart.File) error {
+	// We do these one at a time in the order they were read.
+	for _, obj := range crds {
+		// Read in the resources
+		res, err := i.cfg.KubeClient.Build(bytes.NewBuffer(obj.Data))
+		if err != nil {
+			// We bail out immediately
+			return errors.Wrapf(err, "failed to install CRD %s", obj.Name)
+		}
+		// On dry run, bail here
+		if i.DryRun {
+			i.cfg.Log("WARNING: This chart or one of its subcharts contains CRDs. Rendering may fail or contain inaccuracies.")
+			continue
+		}
+		// Send them to Kube
+		if _, err := i.cfg.KubeClient.Create(res); err != nil {
+			// If the error is CRD already exists, continue.
+			if apierrors.IsAlreadyExists(err) {
+				crdName := res[0].Name
+				i.cfg.Log("CRD %s is already present. Skipping.", crdName)
+				continue
+			}
+			return errors.Wrapf(err, "failed to instal CRD %s", obj.Name)
+		}
+
+		// Invalidate the local cache.
+		if discoveryClient, err := i.cfg.RESTClientGetter.ToDiscoveryClient(); err != nil {
+			// On error, we don't want to bail out, since this is unlikely
+			// to impact the majority of charts.
+			i.cfg.Log("Could not clear the discovery cache: %s", err)
+		} else {
+			i.cfg.Log("Clearing discovery cache")
+			discoveryClient.Invalidate()
```

Comment:
I was looking through this one more time and I think we should do the cache invalidation at the very end, outside of the for loop. When I looked into the code, I think the discovery client calls the api multiple times for each api group, meaning that if someone installs 5-10 CRDs, they will be calling their API hundreds of times

Reply:
Source: https://github.com/kubernetes/cli-runtime/blob/9d670f2021f41f32ed762a6ddbc1c3ac7c441126/pkg/genericclioptions/config_flags.go#L222-L225

Reply:
So originally I was doing this so that the crds folder could have things that depended on each other. And this operates _per file_, not per CRD. But I'll switch it.